scripts/naming/main.go

```diff
@@ -0,0 +1,682 @@
+	if strings.HasPrefix(name, "Nullable") {
+		inner := strings.TrimPrefix(name, "Nullable")
+		normalizedInner := normalizeEntityPart(inner)
+		if normalizedInner != "" {
+			return "Nullable" + normalizedInner
+		}
+		return name
+	}
+
+	if strings.HasSuffix(name, "Paginated") {
+		inner := strings.TrimSuffix(name, "Paginated")
+		normalizedInner := normalizeEntityPart(inner)
+		if normalizedInner != "" {
+			return normalizedInner + "Paginated"
+		}
+		return name
+	}
+
+	normalized := normalizeEntityPart(name)
+	if normalized == "" {
+		return original
+	}
+
+	return normalized
+}
+
+func normalizeEntityPart(part string) string {
+	part = removeDuplicatedNames(part)
+
+	part = removeEmbeddedEntityReferences(part)
+
+	part = strings.Trim(part, "_")
+
+	return part
+}
+
+// This is where most of the work of actually removing "Entity" and "Entities" happens
+func removeEmbeddedEntityReferences(name string) string {
+	if isIdentityOnlySchema(name) {
+		return name
+	}
+
+	result := name
+
+	if !strings.Contains(strings.ToLower(result), "identities") {
+		result = regexp.MustCompile(`Entities([A-Z])`).ReplaceAllString(result, "$1")
+		result = strings.ReplaceAll(result, "Entities_", "_")
+		result = strings.ReplaceAll(result, "_Entities_", "_")
+		result = strings.ReplaceAll(result, "_Entities", "")
+		result = strings.TrimSuffix(result, "Entities")
+	}
+
+	if strings.Contains(strings.ToLower(result), "identity") {
+		// Special handling for identity cases - only remove Entity when it's clearly separate
+		// Handle cases like "ChangeIdentityEntity" and "ChangeIdentityEntityPaginated"
+		if strings.Contains(result, "IdentityEntity") {
+			result = strings.ReplaceAll(result, "IdentityEntity", "Identity")
+		}
+		// Handle underscored cases
+		result = strings.ReplaceAll(result, "Identity_Entity", "Identity")
+		result = strings.ReplaceAll(result, "IdentityEntity_", "Identity_")
+	} else {
+		// Normal entity removal when no identity is present
+		result = regexp.MustCompile(`Entity([A-Z])`).ReplaceAllString(result, "$1")
+		result = strings.ReplaceAll(result, "Entity_", "_")
+		result = strings.ReplaceAll(result, "_Entity_", "_")
+		result = strings.ReplaceAll(result, "_Entity", "")
+		result = strings.TrimSuffix(result, "Entity")
+	}
+
+	// Handle lowercase variants - avoid matching within "identity"
+	if !strings.Contains(strings.ToLower(result), "identity") {
+		result = regexp.MustCompile(`entities([A-Z])`).ReplaceAllString(result, "$1")
+		result = strings.ReplaceAll(result, "entities_", "_")
+		result = strings.ReplaceAll(result, "_entities_", "_")
+		result = strings.ReplaceAll(result, "_entities", "")
+		result = strings.TrimSuffix(result, "entities")
+
+		result = regexp.MustCompile(`entity([A-Z])`).ReplaceAllString(result, "$1")
+		result = strings.ReplaceAll(result, "entity_", "_")
+		result = strings.ReplaceAll(result, "_entity_", "_")
+		result = strings.ReplaceAll(result, "_entity", "")
+		result = strings.TrimSuffix(result, "entity")
```

Comment:
it seems like we have this pattern of logic a few times, would it be worth extracting to a function?

Reply:
good idea, will make a helper function for these